packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -74,123 +90,522 @@ const propTypes = {
+        setInputFocus(event, yearInputRef.current, date.year.length, date.year.length);
+      } else if (event.keyCode === KeyCode.KEY_RIGHT && monthInputRef.current.selectionEnd === date.month.length) {
+        setInputFocus(event, dayInputRef.current, 0, 0);
+      }
+    }
+  };
+
+  const handleYearInputKeydown = (event) => {
+    if (dateFormatOrder === 'MDY') {
+      if ((event.keyCode === KeyCode.KEY_LEFT || event.keyCode === KeyCode.KEY_DELETE || event.keyCode === KeyCode.KEY_BACK_SPACE) && yearInputRef.current.selectionEnd === 0) {
+        setInputFocus(event, dayInputRef.current, date.day.length, date.day.length);
+      }
+    } else if (dateFormatOrder === 'DMY') {
+      if ((event.keyCode === KeyCode.KEY_LEFT || event.keyCode === KeyCode.KEY_DELETE || event.keyCode === KeyCode.KEY_BACK_SPACE) && yearInputRef.current.selectionEnd === 0) {
+        setInputFocus(event, monthInputRef.current, date.month.length, date.month.length);
+      }
+    } else if (dateFormatOrder === 'YMD') {
+      if (event.keyCode === KeyCode.KEY_RIGHT && yearInputRef.current.selectionEnd === date.year.length) {
+        setInputFocus(event, monthInputRef.current, 0, 0);
+      }
+    }
+  };
+
+  const handleOnInputFocus = (event, type) => {
+    if (onFocus) {
+      onFocus(event);
+    }
+
+    setFocused(true);
+
+    if (type === DateUtil.inputType.DAY) {
+      setDayInitialFocused(true);
+      dayInputRef.current.setSelectionRange(0, dayInputRef.current.value.length);
+    } else if (type === DateUtil.inputType.MONTH) {
+      setMonthInitialFocused(true);
+      monthInputRef.current.setSelectionRange(0, monthInputRef.current.value.length);
+    } else {
+      setYearInitialFocused(true);
+      yearInputRef.current.setSelectionRange(0, yearInputRef.current.value.length);
+    }
+  };
+
+  const handleOnInputBlur = (event, type) => {
+    if (onBlur) {
+      onBlur(event);
+    }
+
+    setFocused(false);
+
+    if (type === DateUtil.inputType.DAY) {
+      setDayInitialFocused(false);
+    } else if (type === DateUtil.inputType.MONTH) {
+      setMonthInitialFocused(false);
+    } else {
+      setYearInitialFocused(false);
+    }
+
+    if (type === DateUtil.inputType.DAY || type === DateUtil.inputType.MONTH) {
+      let inputValue = event.target.value;
+
+      // Prepend a 0 to the value when losing focus and the value is single digit.
+      if (inputValue.length === 1) {
+        inputValue = inputValue === '0' ? '' : '0'.concat(inputValue);
+
+        handleDateChange(event, inputValue, type);
+      }
+    }
+  };
+
+  const handleOnButtonClick = (event) => {
+    const attributes = { ...inputAttributes };
+    if (!attributes.readOnly && onCalendarButtonClick && onClick) {
+      onCalendarButtonClick(event, onClick);
+    }
+  };
+
+  const handleOnButtonKeyDown = (event) => {
+    if (onKeyDown) {
+      onKeyDown(event);
+    }
+  };
+
+  const dateInputContainerClasses = cx([
```

Comment:
nit: can we add these where they are used? IMO it's hard to find them and follow there reference in the code when components are this large

Reply:
I have rearranged variables to be defined close to where they are being used now.
Changed here: https://github.com/cerner/terra-framework/pull/1208/commits/ba9d3c9b977ded21c1db0789cbe2b1e43dd3c7c6